// node4_labnode.proto
// Lab Node MCP Server - Student Activity & Analytics
syntax = "proto3";

package node4;

service LabNodeService {
  // Get recent learning concepts
  rpc GetRecentConcepts(GetRecentConceptsRequest) returns (GetRecentConceptsResponse);
  
  // Get concept heatmap
  rpc GetConceptHeatmap(GetConceptHeatmapRequest) returns (GetConceptHeatmapResponse);
  
  // Get weak concepts
  rpc GetWeakConcepts(GetWeakConceptsRequest) returns (GetWeakConceptsResponse);
  
  // Get student activity summary
  rpc GetActivitySummary(GetActivitySummaryRequest) returns (GetActivitySummaryResponse);
  
  // Get class analytics
  rpc GetClassAnalytics(GetClassAnalyticsRequest) returns (GetClassAnalyticsResponse);
}

// ==================== Get Recent Concepts ====================

message GetRecentConceptsRequest {
  string student_id = 1;
  int32 days = 2;
  string curriculum_path = 3;  // Optional
}

message GetRecentConceptsResponse {
  repeated string concepts = 1;
  string student_id = 2;
  int32 period_days = 3;
}

// ==================== Get Concept Heatmap ====================

message GetConceptHeatmapRequest {
  string student_id = 1;
  string curriculum_path = 2;  // Optional
}

message GetConceptHeatmapResponse {
  string student_id = 1;
  map<string, double> heatmap = 2;  // concept -> accuracy (0.0 - 1.0)
  string timestamp = 3;
}

// ==================== Get Weak Concepts ====================

message GetWeakConceptsRequest {
  string student_id = 1;
  double threshold = 2;  // Default: 0.6
  int32 limit = 3;  // Max number of concepts
}

message GetWeakConceptsResponse {
  repeated WeakConcept weak_concepts = 1;
}

message WeakConcept {
  string concept = 1;
  double accuracy = 2;
  int32 attempts = 3;
}

// ==================== Get Activity Summary ====================

message GetActivitySummaryRequest {
  string student_id = 1;
  int32 days = 2;
}

message GetActivitySummaryResponse {
  string student_id = 1;
  int32 period_days = 2;
  int32 total_attempts = 3;
  double average_accuracy = 4;
  int32 concepts_practiced = 5;
  string last_activity = 6;
  string timestamp = 7;
}

// ==================== Get Class Analytics ====================

message GetClassAnalyticsRequest {
  string class_id = 1;
  string start_date = 2;  // Optional
  string end_date = 3;    // Optional
}

message GetClassAnalyticsResponse {
  int32 total_students = 1;
  int32 active_students = 2;
  double average_accuracy = 3;
  int32 at_risk_students = 4;
  repeated CommonWeakConcept common_weak_concepts = 5;
}

message CommonWeakConcept {
  string concept = 1;
  int32 struggling_count = 2;
}
